src/core/exec-invoke.c

```diff
@@ -5332,14 +5376,37 @@ int exec_invoke(
          * different user namespace). */
 
         if (needs_sandboxing && !userns_set_up) {
-                r = setup_private_users(context->private_users, saved_uid, saved_gid, uid, gid,
+                /* If any namespace is delegated with DelegateNamespaces=, always set up a user namespace. */
+                PrivateUsers pu = context->private_users;
+                if (pu == PRIVATE_USERS_NO && context->delegate_namespaces != NAMESPACE_FLAGS_INITIAL)
+                        pu = PRIVATE_USERS_SELF;
```

Comment:
please add a helper for that:

exec_context_get_private_users() or so, that hides this.

i think there are other places to call this, for example exec_context_need_unprivileged_private_users() then